### ISSUE TITLE
Автообновление плагина

### ISSUE THREAD

**Issue by qt-user** [iiko/front.api.sdk#275]
Добрый день. 
Как рекомендуете делать авто обновление плагина? Есть идея качать плагином обновление, вызывать в плагине Dispose()/исключение, запускать процесс/скрипт для копирования файлов и ждать пока хост процесс фронта перезапустит плагин. Или делать это при закрытии фронта.
Ну или правильный способ, тогда вопрос - какие условия/правила обновление через плагин Resto.Front.Api.Updater? Что для этого нужно?

**Comment by Polazhenko** [comment]
Обновить dll плагина налету - сложно, трудно может быть поймать момент, когда Фронт его отпустит, мы предполагаем, что плагин будет сигнализировать о выходе новой версии, пользователь выгрузит фронт и нажмёт скрипт обновления, который скачает дистрибутив и запустит его, дистрибутив будет делать кучу проверок и реализовывать логику обновления с сохранением настроек и т.п.
Возможно, есть какие-то возможности распространения файлов через iiko.chain, сама iiko обновляется при помощи этого инструмента. Но пока не нашли документации как можно было бы воспользоваться этим в своих целях. Заслали запрос на саппорт.

**Comment by Avrmaster** [comment]
Здравствуйте, @Polazhenko!
Удалось ли получить ответ из саппорта?
Тоже интересует данный вопрос.

**Comment by vcpp** [comment]
Да, да, да, автообновление — это востребованная фича, вопросы периодически всплывают (#220), в #111 я даже неосторожно анонсировал эту фичу в 7.2...

Фича действительно реализована в 7.2 и с тех пор допиливается по мере получения обратной связи. Начали с плагинов, разрабатываемых самой компанией iiko, затем расширили на некоторые сторонние плагины, выпускаемые через сборочную инфраструктуру iiko. Цель такого искусственного ограничения — избежать массовых проблем, иметь возможность устранять проблемы в ручном режиме. Как только фича будет готова для широкого применения, напишем новость и инструкцию по использованию.

**Comment by qt-user** [comment]
Мы не стали ждать обновления. Написал свою систему автообновления, простое API на сервере к которому обращается плагин, отправляет свои версию, имя, версию фронта и другие данные. Сервер в ответ генерирует ссылку (под конкретную версию фронта), предоставляет для скачивания Updater.exe и сам архив обновления плагина. Плагин   качает файлы. Функцию обновления положено именно на маленький exe который тянется с сервера до обновления - чтобы на лету можно было менять логику распаковки и тд просто поправив его. Когда архив обновления скачивается успешно во фронте в ватчдоге выводит предложение перезапустить фронт. Плагин в методе Dispose(), при закрытии фронта вызывает процесс Updater.exe если он и сам архив был успешно скачан. Дальше думаю понятно..
Из минусов - фронт должен быть запущен от имени админа. UAC вроде тоже нужно отключать, но его и так всегда отключаем. 
Из плюсов на сервере можно мониторить версии плагинов и данные клиентов - такие как CRMID и онлайн ли он. И кучу разных полезностей например можно передавать ид тимвивера захватив его из реестра (HKEY_LOCAL_MACHINE\SOFTWARE\WOW6432Node\TeamViewer\ClientID). если у вас быстрый доступ в нем

**Comment by lampaa** [comment]
Предлагаю следующий костыль, который работает какое-то время и довольно успешно:

1) Формируем плагин со след. структурой:
```
- plugin/manifest.xml
- plugin/Plugin.dll
- plugin/overwrite.exe
```

2) Файл overwrite.exe просто пытается перезаписать файлы
```
static void Main(string[] args) {
  string folder = args[0];
  string archive = args[1];

  using(ZipFile zip = ZipFile.Read(archive)) {
    var selection = from e in zip.Entries where e.FileName.StartsWith("lib/net472/") select e;

    for (int i = 0; i < 6000; i++) // 1 min
    {
      try {
        selection.ToList().ForEach(entry => {
          string fullPath = folder + "/" + System.IO.Path.GetFileName(entry.FileName);

          File.Delete(fullPath);
          entry.ExtractExistingFile = ExtractExistingFileAction.Throw;
          entry.FileName = System.IO.Path.GetFileName(entry.FileName);
          entry.Extract(folder);
        });

        break;
      } catch (Exception e) {
        Console.WriteLine("BAD write file: " + e.Message);
      }

      Thread.Sleep(10);
    }
  }
}
```

3) Формируем архив следующей структуры:
```
- plugin/manifest.xml
- plugin/Plugin.dll
- plugin/PluginOperations.dll
- тд тп
```

4) Качаем этот архив 
5) Запускаем процесс `overwrite.exe`, который будет пытаться перезаписать заблокированные файлы
6) Вызываем исключение в плагине (но не более 4 исключений за 1 час `sic`!) и у нас появляется окно для перезаписи заблокированных файлов.

**Comment by Kamaew** [comment]
процесс никак еще не автоматизировался в штатном режиме ?
